### DISCUSSION TITLE
API_BASE_URL Setting for Apiserver behind a K8s service.

### DISCUSSION THREAD

**Discussion by Homopatrol** [DependencyTrack/dependency-track#988] [General]
My question is what should I be setting the API_BASE_URL to, given that I am deploying the Community maintained helm chart and the apiserver and frontend are packaged as separate K8s deployments and services within this.
We are trying to login in via the UI and we receive **HTTP 405 Method Not Allowed**. 

We are using this [chart](https://github.com/evryfs/helm-charts/blob/master/charts/dependency-track) and can see from Dependency Track [documentation](https://docs.dependencytrack.org/getting-started/deploy-docker/) and here https://github.com/DependencyTrack/frontend#deployment that something might need to be provided. 

Given that the chart deploys frontend and backend as separate deployments and services, what should we be using as the API_BASE_URL in the frontend so that it can talk to the backend and we don't receive the HTTP error.

### Environment:
  - Dependency-Track Version: 4.2.1
  - Distribution: [ Kubernetes with Helm ]
  - Database Server: [ PostgreSQL ~10.3 ]
  - Browser: Mozilla/5.0

**Comment by namloc2001** [comment]
Updating this with details I posted to the [chart repo issue](https://github.com/evryfs/helm-charts/issues/62) that was raised:

====================

Right, so after staring at this for many hours, I think I understand it and have managed to resolve it within our environment.

I'm afraid I have to disagree with your statement @davidkarlsen 

> that's odd, we run same chart w/o any special setting and it work out of the box. it seems with API_URL="" as today u get the current req url

Because the default setting for the helm chart, regarding Ingress, is that is is disabled, as per [this setting](https://github.com/evryfs/helm-charts/blob/master/charts/dependency-track/values.yaml#L125) and this ["if" statement](https://github.com/evryfs/helm-charts/blob/master/charts/dependency-track/templates/ingress.yaml#L1). Therefore installing with the command in the [README](https://github.com/evryfs/helm-charts/blob/master/charts/dependency-track/README.md), won't work out the box:
```
helm install dependency-track evryfs-oss/dependency-track --namespace dependency-track
```

Instead the `ingress.enabled` settings in [values.yaml](https://github.com/evryfs/helm-charts/blob/master/charts/dependency-track/values.yaml#L125) either needs to be set to `true` or the installation command be updated to be `helm install dependency-track evryfs-oss/dependency-track --set ingress.enabled=true --namespace dependency-track`.

By using the same ingress for both frontend and apiserver, I presume (as we don't want that setup, which I'll explain in a minute, so I haven't tested it) the API_BASE_URL being blank works fine because it does a POST to the API path at itself.

For us this we will be using a slightly different setup as we do not wish to expose any tool interfaces to the Internet unless absolutely required. So we set [API_BASE_URL](https://github.com/evryfs/helm-charts/blob/master/charts/dependency-track/values.yaml#L30) to be `http://127.0.0.1:8081`. Then we install with the command I shared earlier, without exposing the tool via an ingress (or in OpenShift, a route):
```
cd <to_dir_of_local_chart_copy>
helm install dependency-track --set postgresql.serviceAccount.enabled=true --set frontend.replicaCount=1 --set apiserver.resources.requests.cpu="200m" ./
```
And then we use port-forwarding, running these two commands in parallel:
```
oc port-forward svc/<release-name>-frontend 8080:80
oc port-forward svc/<release-name>-apiserver 8081:80
```

We can then browse to 127.0.0.1:8080 to access the UI and it communicates with the apiserver on tcp/8081 via our tunnel.

**Reply by junaid18183** [reply]
Thanks for the comment, saved my day
